include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,622 @@
+   * @param response_code is the response code to send.
+   * @param body is the body of the response.
+   * @param additional_headers are additional headers to send in the response.
+   * @param grpc_status is an optional gRPC status if the connection is a gRPC connection.
+   * @param details are details of any (gRPC) error.
+   */
+  virtual WasmResult sendLocalResponse(uint64_t response_code, string_view body,
+                                       Pairs additional_headers, uint64_t grpc_status,
+                                       string_view details) = 0;
+
+  // Call when the stream closes. See RootInterface.
+  virtual bool onDone() = 0;
+
+  // Call when the stream status has finalized, e.g. for logging. See RootInterface.
+  virtual void onFinalized() = 0;
+
+  // Call just before the Context is deleted. See RootInterface.
+  virtual void onDelete() = 0;
+};
+
+/**
+ * NetworkInterface is the interface between the VM host and the VM for network streams.
+ */
+struct NetworkInterface {
+public:
+  /**
+   * Network
+   * Note: Body data is obtained by the code in the VM via calls implemented by the
+   * proxy-independent host code using the getBuffer() call.
+   */
+
+  /**
+   * Call on a stream Context to indicate that a new network connection has been been created.
+   * Calls onStart().
+   */
+  virtual ProxyAction onNetworkNewConnection() = 0;
+
+  /**
+   * Call on a stream Context to indicate that data has arrived from downstream (e.g. on the
+   * incoming port that was accepted and for which the proxy is acting as a server).
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onDownstreamData() = 0;
+
+  /**
+   * Call on a stream Context to indicate that data has arrived from upstream (e.g. on the
+   * outgoing port that the proxy connected and for which the proxy is acting as a client).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onUpstreamData() = 0;
+
+  /*
+   * The source of the close event.
+   * Unknown is when the source is not known.
+   * Local is when the close was initiated by the proxy.
+   * Remote is when the close was received from the peer.
+   */
+  enum class CloseType : uint32_t {
+    Unknown = 0,
+    Local = 1,
+    Remote = 2,
+  };
+
+  /**
+   * Call on a stream context to indicate that the downstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onDownstreamConnectionClose(CloseType close_type) = 0;
+
+  /**
+   * Call on a stream context to indicate that the upstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onUpstreamConnectionClose(CloseType close_type) = 0;
```

Comment:
How does this actually work for non-terminal network filters? I don't think intermediate filters would necessarily have any knowledge of any upstream connections?